File: backend/src/controllers/addressController.ts
import type { Request, Response } from 'express';
import { getDB } from '../db/database.js';

export const getAddresses = async (req: Request, res: Response) => {
    const db = getDB();
    const userId = (req as any).user.id;

    try {
        const addresses = await db.all('SELECT * FROM addresses WHERE user_id = ? ORDER BY is_default DESC', [userId]);
        res.json(addresses);
    } catch (error) {
        res.status(500).json({ error: 'Failed to get addresses' });
    }
};

export const addAddress = async (req: Request, res: Response) => {
    const db = getDB();
    const userId = (req as any).user.id;
    const { receiver_name, receiver_phone, province, city, district, detail_address, is_default } = req.body;

    try {
        if (is_default) {
            await db.run('UPDATE addresses SET is_default = 0 WHERE user_id = ?', [userId]);
        }

        await db.run(
            `INSERT INTO addresses (user_id, receiver_name, receiver_phone, province, city, district, detail_address, is_default) 
             VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
            [userId, receiver_name, receiver_phone, province, city, district, detail_address, is_default ? 1 : 0]
        );

        res.json({ message: 'Address added successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to add address' });
    }
};

export const updateAddress = async (req: Request, res: Response) => {
    const db = getDB();
    const userId = (req as any).user.id;
    const { id, receiver_name, receiver_phone, province, city, district, detail_address, is_default } = req.body;

    try {
        if (is_default) {
            await db.run('UPDATE addresses SET is_default = 0 WHERE user_id = ?', [userId]);
        }

        await db.run(
            `UPDATE addresses SET receiver_name=?, receiver_phone=?, province=?, city=?, district=?, detail_address=?, is_default=?
             WHERE id = ? AND user_id = ?`,
            [receiver_name, receiver_phone, province, city, district, detail_address, is_default ? 1 : 0, id, userId]
        );

        res.json({ message: 'Address updated successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to update address' });
    }
};

export const deleteAddress = async (req: Request, res: Response) => {
    const db = getDB();
    const userId = (req as any).user.id;
    const { id } = req.params;

    try {
        await db.run('DELETE FROM addresses WHERE id = ? AND user_id = ?', [id, userId]);
        res.json({ message: 'Address deleted successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to delete address' });
    }
};

File: backend/src/main/java/com/diquan/backend/AddressController.java
package com.diquan.backend;

import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/address")
class AddressController {
    private final Sql sql;
    private final AuthService authService;

    AddressController(Sql sql, AuthService authService) {
        this.sql = sql;
        this.authService = authService;
    }

    @GetMapping
    List<Map<String, Object>> list(HttpServletRequest request) {
        AuthenticatedUser user = authService.requireUser(request);
        return sql.list("SELECT * FROM addresses WHERE user_id = ? ORDER BY is_default DESC", user.id());
    }

    @PostMapping
    Map<String, String> add(HttpServletRequest request, @RequestBody Map<String, Object> body) {
        AuthenticatedUser user = authService.requireUser(request);
        if (Values.bool(body, "is_default")) {
            sql.update("UPDATE addresses SET is_default = 0 WHERE user_id = ?", user.id());
        }
        sql.update("""
            INSERT INTO addresses (user_id, receiver_name, receiver_phone, province, city, district, detail_address, is_default)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """,
            user.id(),
            body.get("receiver_name"),
            body.get("receiver_phone"),
            body.get("province"),
            body.get("city"),
            body.get("district"),
            body.get("detail_address"),
            Values.bool(body, "is_default") ? 1 : 0
        );
        return Map.of("message", "Address added successfully");
    }

    @PutMapping
    Map<String, String> update(HttpServletRequest request, @RequestBody Map<String, Object> body) {
        AuthenticatedUser user = authService.requireUser(request);
        if (Values.bool(body, "is_default")) {
            sql.update("UPDATE addresses SET is_default = 0 WHERE user_id = ?", user.id());
        }
        sql.update("""
            UPDATE addresses SET receiver_name=?, receiver_phone=?, province=?, city=?, district=?, detail_address=?, is_default=?
            WHERE id = ? AND user_id = ?
            """,
            body.get("receiver_name"),
            body.get("receiver_phone"),
            body.get("province"),
            body.get("city"),
            body.get("district"),
            body.get("detail_address"),
            Values.bool(body, "is_default") ? 1 : 0,
            body.get("id"),
            user.id()
        );
        return Map.of("message", "Address updated successfully");
    }

    @DeleteMapping("/{id}")
    Map<String, String> delete(HttpServletRequest request, @PathVariable long id) {
        AuthenticatedUser user = authService.requireUser(request);
        sql.update("DELETE FROM addresses WHERE id = ? AND user_id = ?", id, user.id());
        return Map.of("message", "Address deleted successfully");
    }
}
